File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/exception/DuplicateDependencyException.java
package ru.tinkoff.kora.kora.app.annotation.processor.exception;

import com.palantir.javapoet.TypeName;
import ru.tinkoff.kora.annotation.processor.common.ProcessingError;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.kora.app.annotation.processor.component.ComponentDependency;
import ru.tinkoff.kora.kora.app.annotation.processor.component.DependencyClaim;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclaration;

import java.util.List;
import java.util.stream.Collectors;

public class DuplicateDependencyException extends ProcessingErrorException {

    public DuplicateDependencyException(DependencyClaim claim,
                                        ComponentDeclaration declaration,
                                        List<ComponentDeclaration> foundDeclarations) {
        super(List.of(getErrorForDeclarations(claim, declaration, foundDeclarations)));
    }

    public DuplicateDependencyException(List<ComponentDependency.SingleDependency> foundDeclarations,
                                        DependencyClaim claim,
                                        ComponentDeclaration declaration) {
        super(List.of(getErrorForDependencies(claim, declaration, foundDeclarations)));
    }

    private static ProcessingError getErrorForDeclarations(DependencyClaim claim,
                                                           ComponentDeclaration declaration,
                                                           List<ComponentDeclaration> foundDeclarations) {
        var deps = foundDeclarations.stream()
            .map(c -> String.format("- %s", c.declarationString()))
            .collect(Collectors.joining("\n", "Candidates for injection:\n", "")).indent(2);

        return getError(claim, declaration, deps);
    }

    private static ProcessingError getErrorForDependencies(DependencyClaim claim,
                                                           ComponentDeclaration declaration,
                                                           List<ComponentDependency.SingleDependency> foundDeclarations) {
        var deps = foundDeclarations.stream()
            .map(ComponentDependency.SingleDependency::component)
            .map(c -> String.format("- %s", c.declaration().declarationString()))
            .collect(Collectors.joining("\n", "Candidates for injection:\n", "")).indent(2);

        return getError(claim, declaration, deps);
    }

    private static ProcessingError getError(DependencyClaim claim,
                                            ComponentDeclaration declaration,
                                            String deps) {
        if (claim.tag() == null) {
            return new ProcessingError("More than one declaration matches dependency type: " + TypeName.get(claim.type()) + " (no tags)\n"
                                       + deps
                                       + "\nPlease check that injection dependency is declared correctly or that @DefaultComponent annotation is not missing if was intended.",
                declaration.source());
        } else {
            var tagMsg = "@Tag(" + claim.tag() + ".class)";
            return new ProcessingError("More than one declaration matches dependency type: " + TypeName.get(claim.type()) + " with " + tagMsg + "\n"
                                       + deps
                                       + "\nPlease check that injection dependency is declared correctly or that @DefaultComponent annotation is not missing if was intended.",
                declaration.source());
        }
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/exception/DuplicateDependencyException.kt
package ru.tinkoff.kora.kora.app.ksp.exception

import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.kora.app.ksp.component.ComponentDependency
import ru.tinkoff.kora.kora.app.ksp.component.DependencyClaim
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration
import ru.tinkoff.kora.ksp.common.exception.ProcessingError
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.stream.Collectors

data class DuplicateDependencyException(
    val claim: DependencyClaim,
    val declaration: ComponentDeclaration,
    val foundDeclarations: List<ComponentDeclaration>
) : ProcessingErrorException(
    listOf(getErrorForDeclarations(claim, declaration, foundDeclarations))
) {

    constructor(
        foundDeclarations: List<ComponentDependency.SingleDependency>,
        claim: DependencyClaim,
        declaration: ComponentDeclaration
    ) : this(
        claim, declaration, foundDeclarations.map { it.component!!.declaration }.toList()
    )

    companion object {
        private fun getErrorForDeclarations(
            claim: DependencyClaim,
            declaration: ComponentDeclaration,
            foundDeclarations: List<ComponentDeclaration>
        ): ProcessingError {
            val deps = foundDeclarations
                .map { String.format("- %s", it.declarationString()) }
                .joinToString("\n", "Candidates for injection:\n", "").prependIndent("  ")

            return getError(claim, declaration, deps)
        }

        private fun getErrorForDependencies(
            claim: DependencyClaim,
            declaration: ComponentDeclaration,
            foundDeclarations: List<ComponentDependency.SingleDependency>
        ): ProcessingError {
            val deps: String = foundDeclarations
                .map { it.component!!.declaration }
                .map { String.format("- %s", it.declarationString()) }
                .joinToString("\n", "Candidates for injection:\n", "").prependIndent("  ")

            return getError(claim, declaration, deps)
        }

        private fun getError(
            claim: DependencyClaim,
            declaration: ComponentDeclaration,
            deps: String
        ): ProcessingError {
            if (claim.tags.isEmpty()) {
                return ProcessingError(
                    """More than one component matches dependency type: ${claim.type.toTypeName()} (no tags)
                    $deps
                    Please check that injection dependency is declared correctly or that @DefaultComponent annotation is not missing if was intended.""".trimIndent(),
                    declaration.source
                )
            } else {
                val tagMsg: String = claim.tags.stream()
                    .collect(Collectors.joining(", ", "@Tag(", ")"))
                return ProcessingError(
                    """More than one component matches dependency type: ${claim.type.toTypeName()} with $tagMsg
                    $deps
                    Please check that injection dependency is declared correctly or that @DefaultComponent annotation is not missing if was intended.""".trimIndent(),
                    declaration.source
                )
            }
        }
    }
}
